### DISCUSSION TITLE
CheckSum Invalid for Super Famicom

### DISCUSSION THREAD

**Discussion by Crusa1der** [sanni/cartreader#983] [Need help]
Before I get into anything, its important to note that I'm very new to using emulators and I started trying to use it today. If it helps, I'm using a V5 and trying to get the ROM for Super Metroid.

I am able to get through the basic steps; select SNES/SFC, Reads the Card, says its Super Metroid, Read the Cartridge, and then after reading, it says the the Checksum is invalid followed by CRC32.

**Comment by PsychoFox11** [comment]
Have you dumped anything else? Any other SNES? That is, does your OSCR work or is this the first cart you're trying? If it works otherwise, the most common cause is needing to clean the cart. I find the OSCR to be far pickier than real consoles.

**Reply by Crusa1der** [reply]
This is the first cart I've tried. I only have one right now, but should be able to get more soon. I had cleaned the Super Metroid cart throughly until there was no more black stuff coming off of it, then let it dry before using. I also noticed when I put the SD card into my computer, it says the game is there and when I put it into an emulator, it recognized its super metroid but doesn't load.

**Comment by sanni** [comment]
The OSCR is extremely picky, even carts that have been cleaned one or two months ago need to be cleaned again, seems to be especially true for GBA. 
A quick rub with a q-tip dipped in isopropyl alcohol should be sufficient.

Part of the pickiness can be removed by enabling the internal pull-up resistors for the data pins in the setup function of the system. The pull-ups basically stabilize the signals but they can also mask (for developers) interesting stuff like open bus will just become 0xFF. This can be witnesses e.g. with some GBA carts where whole parts of the ROM become 0xFF instead of 0x00.

The other part of the pickiness comes from the fact that almost none of the timings have been adjusted using a logic analyzer, so the OSCR does not behave like a real console and might be a couple us off the optimal timing. 
This is especially the case with the SNES code which doesn't even toggle the RD signal for every byte and just keeps it enabled all the time, reason being it was the first system we added so it's the oldest code base with the most inaccuracies.

![timing_snes](https://github.com/sanni/cartreader/assets/19169494/da53aea2-772e-4ff2-a227-04801017fe46)

And now it is more of a if-it-works-don't-change-it situation because changing it would mean that a lot of the stuff would need to be retested. But maybe some day. 😄
